libs/ml/examples/boston_housing_distributed_learning/main.cpp

```diff
@@ -0,0 +1,180 @@
+using TensorType       = fetch::math::Tensor<DataType>;
+using VectorTensorType = std::vector<TensorType>;
+using SizeType         = fetch::math::SizeType;
+
+std::shared_ptr<TrainingClient<TensorType>> MakeClient(std::string const &     id,
+                                                       ClientParams<DataType> &client_params,
+                                                       std::string const &     data,
+                                                       std::string const &     labels,
+                                                       float                   test_set_ratio)
+{
+  // Initialise model
+  std::shared_ptr<fetch::ml::Graph<TensorType>> g_ptr =
+      std::make_shared<fetch::ml::Graph<TensorType>>();
+
+  client_params.inputs_names = {g_ptr->template AddNode<PlaceHolder<TensorType>>("Input", {})};
+  g_ptr->template AddNode<FullyConnected<TensorType>>("FC1", {"Input"}, 13u, 10u);
+  g_ptr->template AddNode<Relu<TensorType>>("Relu1", {"FC1"});
+  g_ptr->template AddNode<FullyConnected<TensorType>>("FC2", {"Relu1"}, 10u, 10u);
+  g_ptr->template AddNode<Relu<TensorType>>("Relu2", {"FC2"});
+  g_ptr->template AddNode<FullyConnected<TensorType>>("FC3", {"Relu2"}, 10u, 1u);
+  client_params.label_name = g_ptr->template AddNode<PlaceHolder<TensorType>>("Label", {});
+  client_params.error_name =
+      g_ptr->template AddNode<MeanSquareErrorLoss<TensorType>>("Error", {"FC3", "Label"});
+
+  // Initialise DataLoader
+  std::shared_ptr<fetch::ml::dataloaders::TensorDataLoader<TensorType, TensorType>> dataloader_ptr =
+      std::make_shared<fetch::ml::dataloaders::TensorDataLoader<TensorType, TensorType>>();
+  TensorType data_tensor  = fetch::ml::dataloaders::ReadCSV<TensorType>(data);
+  TensorType label_tensor = fetch::ml::dataloaders::ReadCSV<TensorType>(labels);
+  dataloader_ptr->AddData(data_tensor, label_tensor);
+
+  dataloader_ptr->SetTestRatio(test_set_ratio);
+  dataloader_ptr->SetRandomMode(true);
+  // Initialise Optimiser
+  std::shared_ptr<fetch::ml::optimisers::Optimiser<TensorType>> optimiser_ptr =
+      std::make_shared<fetch::ml::optimisers::AdamOptimiser<TensorType>>(
+          std::shared_ptr<fetch::ml::Graph<TensorType>>(g_ptr), client_params.inputs_names,
+          client_params.label_name, client_params.error_name, client_params.learning_rate);
+
+  return std::make_shared<TrainingClient<TensorType>>(id, g_ptr, dataloader_ptr, optimiser_ptr,
+                                                      client_params);
+}
+
+int main(int ac, char **av)
+{
+  if (ac < 3)
+  {
+    std::cout << "Usage : " << av[0] << " PATH/TO/boston_data.csv PATH/TO/boston_label.csv"
+              << std::endl;
+    return 1;
+  }
+
+  CoordinatorParams      coord_params;
+  ClientParams<DataType> client_params;
+
+  SizeType number_of_clients    = 10;
+  SizeType number_of_rounds     = 100;
+  coord_params.mode             = CoordinatorMode::SEMI_SYNCHRONOUS;
```

Comment:
before we put in this PR; can you try changing this to ASYNCHRONOUS, and running the example for 10 epochs with perhaps 4 clients (split the data into 4 groups)? And then lets take a look at plots of loss for all 4 clients.
Also add a test/evaluate method which shows performance on the full data set? and we can see what that looks like for all 4 clients as well.